### PR TITLE
Client API Integration

### PR DESCRIPTION
**Pull request by mauesrog**
**Problem:**
Missing support for IdToken Hints.

**Solution:**
- Add new TokenType for IdToken Hints.
- Augment IdTokenAttestationModel with optional trusted issuers.

**Validation:**
Unit tests and tested end-to-end.

**Type of change:**
- [x] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

**Risk**:
- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [x] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [ ] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

### CODE REVIEW

File: lib/verifiable_credential/ClaimToken.ts [hunk 1]
```diff
@@ -165,6 +170,8 @@ export default class ClaimToken {
           return new ClaimToken(TokenType.siopPresentationExchange, <string>token, id);
         } else if (payload.attestations) {
           return new ClaimToken(TokenType.siopPresentationAttestation, <string>token, id);
+        } else if (id === VerifiableCredentialConstants.TOKEN_SI_ISS) {
```
Comment (gproanomsft) [inline lib/verifiable_credential/ClaimToken.ts:173]: is the idTokenHint not in payload.attestations?

File: tests/ClaimToken.spec.ts [hunk 2]
```diff
@@ -43,4 +54,48 @@ import { ValidationError } from '../lib';
...
+  describe('create()', () => {
+    const iss = VerifiableCredentialConstants.TOKEN_SI_ISS;
+    let getTokenPayloadSpy: jasmine.Spy<(token: string) => any>;
+
+    beforeAll(() => {
+      getTokenPayloadSpy = spyOn(<any>ClaimToken, 'getTokenPayload').withArgs(sampleToken);
+    });
+
+    beforeEach(() => {
+      decodeSpy.and.stub();
```
Comment (gproanomsft) [inline tests/ClaimToken.spec.ts:67]: why are you stubbing this out?

File: tests/ClaimToken.spec.ts [hunk 3]
```diff
@@ -43,4 +54,48 @@ import { ValidationError } from '../lib';
...
+    });
+
+    it('should create IdTokenHint type tokens', () => {
+      // Mock payload creation.
+      const payload = { iss };
+      getTokenPayloadSpy.and.returnValue(payload);
```
Comment (gproanomsft) [inline tests/ClaimToken.spec.ts:73]: core functionality of ClaimToken is handling tokens, overreliance on spies that is not ok here

File: tests/Validator.spec.ts [hunk 4]
```diff
@@ -377,4 +374,42 @@ describe('Validator', () => {
...
     expect(response.result).toBeFalsy();
   });
+
+  describe('setValidationResult()', () => {
+    let validator: Validator;
+    let decodeSpy: jasmine.Spy<() => void>;
+
+    beforeAll(() => {
+      validator = new Validator(new ValidatorBuilder(new Crypto(new CryptoBuilder())));
+      decodeSpy = spyOn((<any>ClaimToken).prototype, 'decode');
```
Comment (gproanomsft) [inline tests/Validator.spec.ts:384]: i dont think there's a reason to decode here, there should be lots of unit test code here showing how to create a token

**Review by gproanomsft — CHANGES_REQUESTED** [review]

File: lib/verifiable_credential/ClaimToken.ts (hunk 1 above)
Comment (mauesrog) [reply lib/verifiable_credential/ClaimToken.ts:173]: It is. But this code is getting called on a deeper level as well. First the issuance SiopID Token, then in the IdToken Hint itself. This code is just for the latter case.

File: tests/ClaimToken.spec.ts (hunk 3 above)
Comment (mauesrog) [reply tests/ClaimToken.spec.ts:73]: There are multiple tests testing the real thing. This is not about ClaimToken functionality. I'm taking Ali's feedback here. Unit tests are meant to tests units, we have to assume the preexisting units are tested. I'm testing the new units and nothing else. End-to-end tests are not the goal I was trying to achieve here.

File: tests/Validator.spec.ts (hunk 4 above)
Comment (mauesrog) [reply tests/Validator.spec.ts:384]: ClaimToken calls decode by default, which is not what we're testing. There other ways to unit test ClaimToken are there already, no need to apply an overkill apporach to a completely unrelated unit.

File: tests/ClaimToken.spec.ts (hunk 2 above)
Comment (mauesrog) [reply tests/ClaimToken.spec.ts:67]: See other comment.

File: tests/ClaimToken.spec.ts (hunk 3 above)
Comment (gproanomsft) [reply tests/ClaimToken.spec.ts:73]: we test by sending inputs to functions and validating outputs.  let's mock things we can't reasonably test, saying that the inputs of the function don't matter because a subset of them might be tested elsewhere does not test the function

case in point, i suggest that aud of the token hint is an array, that is not tested

don't over rely on spies and hide bugs to avoid redundancy

Comment (mauesrog) [reply tests/ClaimToken.spec.ts:73]: It's not an overreliance on spies but a unit test design principle. f(g(h(a))). If i'm unit testing h, I will mock f and g.

Comment (mauesrog) [reply tests/ClaimToken.spec.ts:73]: The design principle has to do with following this pattern not with how difficult or easy it is to not use spies. I'm not using spies because it's easier but because I'm trying to adopt Ali's unit testing principle here.

Comment (mauesrog) [reply tests/ClaimToken.spec.ts:73]: One great thing about this design is that when you change the unit (a method, for example), you only need to update a single unit test section instead of an arbitrarily large suites of unit tests that are sometimes puzzling to decipher since they're not really about the unit you changed at all.

Comment (gproanomsft) [reply tests/ClaimToken.spec.ts:73]: there's an awfully big fallacy here that code you stub out is fully tested.  i don't think we have any code that we can have confidence in and it is reflected downstream in our numerous issues

what i'm reading here is "because we don't reuse code well in unit tests, we should strive to minimize our testing efforts"

lots of different ways to test things and you can look at the geneva-common test cases on how things like token generation are handled

but the design principle of "the input doesnt matter because we can stub away" does not pass muster

Comment (mauesrog) [reply tests/ClaimToken.spec.ts:73]: I don't think it's a fallacy when it is an assumption I'm admitting to make. The scope of this PR is not to fully unit test ClaimToken but to address only the changes that were introduced by this PR.

I was purposely surgical in the changes I made to ClaimToken:

1) Add a new TokenType.
2) Return an IdTokenHint ClaimToken when the SIOP Id Token key matches 'https://self-issued.me'.
3) Pass in token key to `create` method.

The tests I added are equally surgical. All these three scenarios are tested. I don't get why you're suggesting to test anything outside these scenarios. It is not within the scope of this item.

File: lib/rules_model/TrustedIssuerModel.ts [hunk 5]
```diff
@@ -11,14 +13,19 @@ export class TrustedIssuerModel {
...
   /**
    * Populate an instance of TrustedIssuerModel from any instance
    * @param input object instance to populate from
    */
   populateFrom (input: any): void {
-    this.iss = input.iss;
+    const { iss } = input;
+
+    if (!iss) {
+      throw new RulesValidationError('Trusted issuer requires a valid DID for property \'iss\'.');
```
Comment (gproanomsft) [inline lib/rules_model/TrustedIssuerModel.ts:26]: if your doing this, why not type check?

File: tests/ClaimToken.spec.ts (hunk 3 above)
Comment (gproanomsft) [reply tests/ClaimToken.spec.ts:73]: when we write tests with the assumption that things work we miss the point of proving things work

your tests assume that the only thing that decode can give you is perfect path, so you lose sight of testing inputs and expecting outputs

our overreliance of spies helps create the conditions to have non functional code that has 100% code coverage and 100% passing rate

redundancy and amplification of issues compared to shipping non functional code 

here is a perfectly valid id token hint that your design fails to test
 
      const ct = ClaimToken.create(       'eyJhbGciOiJFUzI1NksiLCJraWQiOiJkaWQ6aW9uOkVpRDJRWXNLUmh6QU1wU3VhMDhvSk5hSkRVcUdDQkJvSGVnXzk4VW8xd0lFZ2c6ZXlKa1pXeDBZU0k2ZXlKd1lYUmphR1Z6SWpwYmV5SmhZM1JwYjI0aU9pSnlaWEJzWVdObElpd2laRzlqZFcxbGJuUWlPbnNpY0hWaWJHbGpTMlY1Y3lJNlczc2lhV1FpT2lKemFXY2lMQ0p3ZFdKc2FXTkxaWGxLZDJzaU9uc2lZM0oySWpvaWMyVmpjREkxTm1zeElpd2lhM1I1SWpvaVJVTWlMQ0o0SWpvaU9GcEllRXB1Y0hkR1NFUlJTRXhJUmxwbGFXOVJSakk0ZW1aUGVHbEdURjh5TUZSYU5rMUVkMDVJUVNJc0lua2lPaUkyZDIxc1JqTTJOMjFvUlVGamQzQjZZMUE1VURkbFJWQXlTMlpxTTBGeVp6Y3lVR3d4VG5SWlFtOUpJbjBzSW5CMWNuQnZjMlZ6SWpwYkltRjFkR2hsYm5ScFkyRjBhVzl1SWwwc0luUjVjR1VpT2lKRlkyUnpZVk5sWTNBeU5UWnJNVlpsY21sbWFXTmhkR2x2Ymt0bGVUSXdNVGtpZlYxOWZWMHNJblZ3WkdGMFpVTnZiVzFwZEcxbGJuUWlPaUpGYVVOWlJrNVVkMFpxVW1GWlZFNVpYM2RIVnpJelVHZE1NSE5aYm5FNFZWOXhkVkk1ZFRCc2VXSTFMV3RuSW4wc0luTjFabVpwZUVSaGRHRWlPbnNpWkdWc2RHRklZWE5vSWpvaVJXbENRelI0U3poUWQyRXRjbFJMVVdRMFgyZFNVV1k0WDI5QlZtOUxlVzVLYmtveGFFOVVTa1pPZFVSMVFTSXNJbkpsWTI5MlpYSjVRMjl0YldsMGJXVnVkQ0k2SWtWcFJGbFhWREJWZFhaRE5FNUxhVGRmYm1ObVpVcFhTM1pQUkZaMWRtMHdZa3BCV0ZGUFFtcGlRMU5RWkZFaWZYMCNzaWciLCJ0eXAiOiJKV1QifQ.eyJhdHRlc3RhdGlvbnMiOiJ0ZXN0cyBhcmUgdGhlcmUgdG8gcHJvdmUgdGhpbmdzIHdvcmsiLCJqdGkiOiIyYjk2ZjBmMDVjMWM0OWI3OGQwZDQzYmFiNGI5NzlmNSIsImRpZCI6ImRpZDppb246RWlEMlFZc0tSaHpBTXBTdWEwOG9KTmFKRFVxR0NCQm9IZWdfOThVbzF3SUVnZzpleUprWld4MFlTSTZleUp3WVhSamFHVnpJanBiZXlKaFkzUnBiMjRpT2lKeVpYQnNZV05sSWl3aVpHOWpkVzFsYm5RaU9uc2ljSFZpYkdsalMyVjVjeUk2VzNzaWFXUWlPaUp6YVdjaUxDSndkV0pzYVdOTFpYbEtkMnNpT25zaVkzSjJJam9pYzJWamNESTFObXN4SWl3aWEzUjVJam9pUlVNaUxDSjRJam9pT0ZwSWVFcHVjSGRHU0VSUlNFeElSbHBsYVc5UlJqSTRlbVpQZUdsR1RGOHlNRlJhTmsxRWQwNUlRU0lzSW5raU9pSTJkMjFzUmpNMk4yMW9SVUZqZDNCNlkxQTVVRGRsUlZBeVMyWnFNMEZ5WnpjeVVHd3hUblJaUW05SkluMHNJbkIxY25CdmMyVnpJanBiSW1GMWRHaGxiblJwWTJGMGFXOXVJbDBzSW5SNWNHVWlPaUpGWTJSellWTmxZM0F5TlRack1WWmxjbWxtYVdOaGRHbHZia3RsZVRJd01Ua2lmVjE5ZlYwc0luVndaR0YwWlVOdmJXMXBkRzFsYm5RaU9pSkZhVU5aUms1VWQwWnFVbUZaVkU1WlgzZEhWekl6VUdkTU1ITlpibkU0VlY5eGRWSTVkVEJzZVdJMUxXdG5JbjBzSW5OMVptWnBlRVJoZEdFaU9uc2laR1ZzZEdGSVlYTm9Jam9pUldsQ1F6UjRTemhRZDJFdGNsUkxVV1EwWDJkU1VXWTRYMjlCVm05TGVXNUtia294YUU5VVNrWk9kVVIxUVNJc0luSmxZMjkyWlhKNVEyOXRiV2wwYldWdWRDSTZJa1ZwUkZsWFZEQlZkWFpETkU1TGFUZGZibU5tWlVwWFMzWlBSRloxZG0wd1lrcEJXRkZQUW1waVExTlFaRkVpZlgwIiwic3ViX2p3ayI6eyJhbGciOiJFUzI1NksiLCJjcnYiOiJzZWNwMjU2azEiLCJraWQiOiJkaWQ6aW9uOkVpRDJRWXNLUmh6QU1wU3VhMDhvSk5hSkRVcUdDQkJvSGVnXzk4VW8xd0lFZ2c6ZXlKa1pXeDBZU0k2ZXlKd1lYUmphR1Z6SWpwYmV5SmhZM1JwYjI0aU9pSnlaWEJzWVdObElpd2laRzlqZFcxbGJuUWlPbnNpY0hWaWJHbGpTMlY1Y3lJNlczc2lhV1FpT2lKemFXY2lMQ0p3ZFdKc2FXTkxaWGxLZDJzaU9uc2lZM0oySWpvaWMyVmpjREkxTm1zeElpd2lhM1I1SWpvaVJVTWlMQ0o0SWpvaU9GcEllRXB1Y0hkR1NFUlJTRXhJUmxwbGFXOVJSakk0ZW1aUGVHbEdURjh5TUZSYU5rMUVkMDVJUVNJc0lua2lPaUkyZDIxc1JqTTJOMjFvUlVGamQzQjZZMUE1VURkbFJWQXlTMlpxTTBGeVp6Y3lVR3d4VG5SWlFtOUpJbjBzSW5CMWNuQnZjMlZ6SWpwYkltRjFkR2hsYm5ScFkyRjBhVzl1SWwwc0luUjVjR1VpT2lKRlkyUnpZVk5sWTNBeU5UWnJNVlpsY21sbWFXTmhkR2x2Ymt0bGVUSXdNVGtpZlYxOWZWMHNJblZ3WkdGMFpVTnZiVzFwZEcxbGJuUWlPaUpGYVVOWlJrNVVkMFpxVW1GWlZFNVpYM2RIVnpJelVHZE1NSE5aYm5FNFZWOXhkVkk1ZFRCc2VXSTFMV3RuSW4wc0luTjFabVpwZUVSaGRHRWlPbnNpWkdWc2RHRklZWE5vSWpvaVJXbENRelI0U3poUWQyRXRjbFJMVVdRMFgyZFNVV1k0WDI5QlZtOUxlVzVLYmtveGFFOVVTa1pPZFVSMVFTSXNJbkpsWTI5MlpYSjVRMjl0YldsMGJXVnVkQ0k2SWtWcFJGbFhWREJWZFhaRE5FNUxhVGRmYm1ObVpVcFhTM1pQUkZaMWRtMHdZa3BCV0ZGUFFtcGlRMU5RWkZFaWZYMCNzaWciLCJrdHkiOiJFQyIsIngiOiI4Wkh4Sm5wd0ZIRFFITEhGWmVpb1FGMjh6Zk94aUZMXzIwVFo2TUR3TkhBIiwieSI6IjZ3bWxGMzY3bWhFQWN3cHpjUDlQN2VFUDJLZmozQXJnNzJQbDFOdFlCb0kifSwiYXVkIjoiZG9lc24ndCBtYXR0ZXIiLCJpc3MiOiJodHRwczovL3NlbGYtaXNzdWVkLm1lIiwic3ViIjoiUlZkS29OemxJT0R1enYxSjJ6cEl5R3FUc01Ia1Blc09CRGJXeDlrT0NGayIsImlhdCI6MTYxNjczOTUzOSwiZXhwIjoxNjE3MzQ0MzM5fQ.XmfXvxSPM1f-CyEHi9Og9SrfRGZRmLIABninPORgjOOT4YhjiaLnhRla10yG2ha-A8zmgELIuypz3-tBFK6PUw',
        'https://self-issued.me');

**Review by gproanomsft — CHANGES_REQUESTED** [review]

File: tests/ClaimToken.spec.ts (hunk 2 above)
Comment (gproanomsft) [reply tests/ClaimToken.spec.ts:67]: spy things that can't be done

Comment (mauesrog) [reply tests/ClaimToken.spec.ts:67]: Disagree. That's what integration tests are for. Unit tests test units.

File: tests/ClaimToken.spec.ts (hunk 3 above)
Comment (mauesrog) [reply tests/ClaimToken.spec.ts:73]: Agree with all those assumptions, disagree with blaming spies for100% coverage/test rate. My test is meant to do a surgical task and that's what it's doing and nothing more. Do we need more tests for decode and the other methods in this class? Sure. Is this PR the place to do it? No.

Comment (mauesrog) [reply tests/ClaimToken.spec.ts:73]: I even volunteer to create a bug and open a PR addressing all of this. I simply don't see this PR as the place for that.

Comment (gproanomsft) [reply tests/ClaimToken.spec.ts:73]: in continuing to insist on using spies, you are missing the point that you have not adequately tested your surgical changes

**Review by gproanomsft — CHANGES_REQUESTED** [review]

File: tests/ClaimToken.spec.ts (hunk 3 above)
Comment (PreetiRastogi) [reply tests/ClaimToken.spec.ts:73]: Thanks Guillermo for a thorough feedback. I was going through the comments above. For me unit testing is to validate the functionality of a section of the code and component tests are for more for complete module testing, and integration tests for end-end testing. With this regard, I agree with Mau's changes as he covers the testing around the additional claimtoken type, that he is adding the support for it in this PR.

Adding @mualigit as fyi, would love to get his inputs as well.

**Review by PreetiRastogi — DISMISSED** [review]

File: lib/verifiable_credential/ClaimToken.ts [hunk 6]
```diff
@@ -164,14 +164,14 @@ export default class ClaimToken {
       // compact jwt      
       // Check type of token
       if (payload.iss === VerifiableCredentialConstants.TOKEN_SI_ISS) {
-        if (payload.contract) {
+        if (id === VerifiableCredentialConstants.TOKEN_SI_ISS) {
```
Comment (gproanomsft) [inline lib/verifiable_credential/ClaimToken.ts:167]: i think it'd be cleaner to remove the else statements since we return from inside the if block

File: lib/verifiable_credential/ClaimToken.ts [hunk 7]
```diff
@@ -202,7 +209,7 @@ export default class ClaimToken {
       }
       else {
         for (let tokenKey in token) {
-          const claimToken = ClaimToken.create(token[tokenKey]);
+          const claimToken = ClaimToken.create(token[tokenKey], tokenKey);
```
Comment (gproanomsft) [inline lib/verifiable_credential/ClaimToken.ts:212]: i'm curious about why this change?  and if there's some path we haven't thought about

**Review by gproanomsft — CHANGES_REQUESTED** [review]

File: lib/verifiable_credential/ClaimToken.ts (hunk 6 above)
Comment (mauesrog) [reply lib/verifiable_credential/ClaimToken.ts:167]: Yeah, 100% agree. Just didn't want to change unnecessary stuff.

File: lib/verifiable_credential/ClaimToken.ts (hunk 7 above)
Comment (mauesrog) [reply lib/verifiable_credential/ClaimToken.ts:212]: I agree that it might have unexpected paths, I'm not sure. We need it in order to recognize that it's a SIOP Id Token passed in as an Id Token Hint (line 167 in ClaimToken).

Comment (gproanomsft) [reply lib/verifiable_credential/ClaimToken.ts:218]: line 167 returns, so that path can't come here.  something doesn't quite add up

**Review by gproanomsft — COMMENTED** [review]
pending test deployment and validation

File: lib/verifiable_credential/ClaimToken.ts (hunk 7 above)
Comment (mauesrog) [reply lib/verifiable_credential/ClaimToken.ts:218]: It's the other way round. This method calls create which like you said returns at line 167. I'm not sure what you mean?

Comment (mauesrog) [reply lib/verifiable_credential/ClaimToken.ts:218]: To be quite clear, line 167 only returns _dure_ to passing in the id here.

Comment (mauesrog) [reply lib/verifiable_credential/ClaimToken.ts:218]: In other words, `id` in line 167 is `tokenKey` from here.

Comment (gproanomsft) [reply lib/verifiable_credential/ClaimToken.ts:218]: yeah the UX makes it look like they're in teh same function

**Review by gproanomsft — APPROVED** [review]
